docs/design/hierarchical-queue-on-capacity-plugin.md

```diff
@@ -0,0 +1,133 @@
+
+## Design detail
+
+### Data structure modifications
+
+- Extend the queueAttr struct in capacity plugin to include two new fields:
+  - `parents`: Store all the parent queues of the current queue, including the root queue.
+  - `children`: Store all the child queues of the current queue.
+
+  ```go
+  type queueAttr struct {
+  	queueID api.QueueID
+  	name    string
+  	share   float64
+  	// parents represents the parents of the queue, including the root queue
+  	parents []api.QueueID
+
+  	deserved  *api.Resource
+  	allocated *api.Resource
+  	request   *api.Resource
+  	// elastic represents the sum of job's elastic resource, job's elastic = job.allocated - job.minAvailable
+  	elastic *api.Resource
+  	// inqueue represents the resource request of the inqueue job
+  	inqueue    *api.Resource
+  	capability *api.Resource
+  	// realCapability represents the resource limit of the queue, LessEqual capability
+  	realCapability *api.Resource
+  	guarantee      *api.Resource
+  	// children represents the child queues of the queue
+  	children map[api.QueueID]*queueAttr
+  }
+  ```
+
+- Extend the capacityPlugin struct in capacity plugin to include one new field:
+  - `rootQueue`: Represents the root queue. It is the parent of all queues.
+  
+  ```go
+  type capacityPlugin struct {
+    totalResource  *api.Resource
+    totalGuarantee *api.Resource
+    rootQueue api.QueueID
+    
+    queueOpts map[api.QueueID]*queueAttr
+    // Arguments given for the plugin
+    pluginArguments framework.Arguments
+  }
+  ```
+
+**Note:** When the scheduler has the capacity plugin enabled and `EnabledHierarchy` is set to true, a root queue with queueID `root` will be automatically created. In the `OpenSession` function, the `Deserved` and `Capability` fields of the root queue are updated each time based on `ssn.TotalResource`.
+
+### Key function modifications
+
+- `OnSessionOpen`: Check that the sum of the deserved/guarantee values of each queue's child queues does not exceed the deserved/guarantee of the parent queue, and that the capability of any child queue does not exceed that of the parent queue. 
+  - If the check reveals non-compliance with the aforementioned conditions, an error message should be printed, and the relevant functions for the session should return a disallowed status. 
+  - For example, in the case of `AddJobEnqueueableFn`, it should directly return `util.Reject`, preventing the job from being enqueued. No scheduling should be performed in situations that do not align with the previously described conditions. Users need to adjust the queues based on the printed error information until normal scheduling can be resumed.
+- `AddQueueOrderFn`: Prioritize leaf queues (queues without child queues) when considering queue ordering. For two leaf queues, the ordering will be determined by considering their parent queues layer by layer, starting from the root queue.
```

Comment:
There are many limitation and validation in hierarchical queue, some of the validation mentioned in the document are in plugins, such as AddQueueOrderFn here, and some are in webhooks, such as limitations part below. It is best to perform validation in a unified place. You need to clarify where to put it.